### ISSUE TITLE
Utilizing FFTransformer framework for binary classification but the results always showing the same class

### ISSUE THREAD

**Issue by YvooneYIN** [yandex-research/rtdl-revisiting-models#46]
Hi! I am really appreciate for this incredible work of FTTransformer, and I am wondering how I could utilize this for a simple binary classification problem on a tabular data.

First, I generated some mock data and binary labels through **sklearn.datasets.make_classification** 
`X_m1,y_m1 = make_classification(n_samples=5000, n_features=455, n_informative=455, n_redundant=0, n_repeated=0,n_classes=2, n_clusters_per_class=2,weights=None,flip_y=0, class_sep=1.0, hypercube=True, shift=0.0, scale=1.0,shuffle=True, random_state=100)`

And then I fed these data into FTTransformer through these code:
`basemodel = FTTransformer( n_cont_features=455,cat_cardinalities[],d_block=4,n_blocks=2,attention_n_heads=2,attention_dropout=0.0,ffn_dropout=0.0,residual_dropout=0.0,d_out=1,ffn_d_hidden_multiplier = 2.0)`

A simple BCEloss function and SGD optimizer were defined using 
`loss_function = nn.BCELoss()`
`optimizer = SGD(basemodel.parameters(),lr=0.00001)`

The data was preprocessed using:
`X_m1 = torch.tensor(X_m1,dtype=torch.float32)`
`y_m1 = torch.tensor(y_m1,dtype=torch.float32).unsqueeze(1)`

At last, I want to see the results of the train data (I've checked the results of validation data initially, yet they behaved weird, so I turn to the train data to make sure if the model was actually learning...)
my code of the training process is:
```
    n_epochs = 10
    for epoch in range(n_epochs):
    basemodel.train()
    total_loss = 0
    all_labels = []
    all_predictions = []

    for i in range(len(X_m1)):  

        optimizer.zero_grad()
        outputs = basemodel(X_m1[i].unsqueeze(0),x_cat=None)
        loss = loss_function(torch.sigmoid(outputs), y_m1[i].unsqueeze(1))
        loss.backward()
        optimizer.step()

        total_loss += loss.item()
        predicted_prob = torch.sigmoid(outputs)
        predicted = (predicted_prob >= 0.5).float()

        all_labels.extend(y_m1[i].unsqueeze(1).view(-1).cpu().numpy())
        all_predictions.extend(predicted_prob.view(-1).cpu().detach().numpy())
    
  
    avg_loss = total_loss / len(y_m1)
    all_labels = np.array(all_labels)
    all_predictions = np.array(all_predictions)
    all_predictions_binary = (all_predictions >= 0.5).astype(int)  

    tn, fp, fn, tp = confusion_matrix(all_labels, all_predictions_binary).ravel()
    fpr = fp / (fp + tn) if (fp + tn) > 0 else 0
    fnr = fn / (fn + tp) if (fn + tp) > 0 else 0
    accuracy = (tp + tn) / (tp + fp + fn + tn) if (tp + fp + fn + tn) > 0 else 0
    auc = roc_auc_score(all_labels, all_predictions)
    
    print(f"Epoch {epoch+1}/{n_epochs}, Loss: {avg_loss:.4f}, Accuracy: {accuracy:.4f}, AUC: {auc:.4f}, FPR: {fpr:.4f}, FNR: {fnr:.4f}")
```
And therefore, the output was:

Epoch 1/10, Loss: 0.7056, Accuracy: 0.5000, AUC: 0.5013, FPR: 0.0000, FNR: 1.0000
Epoch 2/10, Loss: 0.7042, Accuracy: 0.5000, AUC: 0.5007, FPR: 0.0000, FNR: 1.0000
Epoch 3/10, Loss: 0.7030, Accuracy: 0.5000, AUC: 0.5002, FPR: 0.0000, FNR: 1.0000
Epoch 4/10, Loss: 0.7019, Accuracy: 0.5000, AUC: 0.4997, FPR: 0.0000, FNR: 1.0000
Epoch 5/10, Loss: 0.7009, Accuracy: 0.5000, AUC: 0.4990, FPR: 0.0000, FNR: 1.0000
Epoch 6/10, Loss: 0.7001, Accuracy: 0.5000, AUC: 0.4984, FPR: 0.0000, FNR: 1.0000
Epoch 7/10, Loss: 0.6993, Accuracy: 0.5000, AUC: 0.4977, FPR: 0.0000, FNR: 1.0000
Epoch 8/10, Loss: 0.6986, Accuracy: 0.5000, AUC: 0.4971, FPR: 0.0000, FNR: 1.0000
Epoch 9/10, Loss: 0.6981, Accuracy: 0.5000, AUC: 0.4964, FPR: 0.0000, FNR: 1.0000
Epoch 10/10, Loss: 0.6975, Accuracy: 0.5000, AUC: 0.4956, FPR: 0.0000, FNR: 1.0000

And I tried to print for the prediction label (binary label but not the original logit), they were exactly all 0 or 1

What I have tried includes:
- Standardize the data using **StandardScaler**
- Changing different threshold instead of 0.5 in the _sigmoid(logit) transforming to predicted label_ process, using cross validation to choose the best threshold which leads to the highest f1 score
- Various learning rate, optimizer
- Different parameters of the FTTransformer, including d_block, n_blocks, attention_n_heads, attention_dropout, ffn_dropout, residual_dropout
- More/less epoch and larger/smaller batch sizes
- using a batch of samples in dataloader instead of a single sample when training

However, above adjustments all seems not working, the model still predict all the samples into a single class, either 1 or 0 ... 

I have tried a simple neural network with the same data, and it worked fine.

So, I am wondering if there are some kindly suggestions or insights into this problem? Why it seems that my practice with FTTransformer doesn't learn anything? Thanks for any suggestions!

**Comment by Yura52** [comment]
Hi! Have you seen our [official usage example](https://github.com/yandex-research/rtdl-revisiting-models/blob/main/package/example.ipynb)? I recommend trying to reproduce it without any changes. If it runs successfully, then it can be a good starting point for debugging your code. For example, I see that your code uses `sigmoid` + `BCELoss`, which is less numerically stable than using `BCEWithLogitsLoss`, and there is a bunch of squeeze/unsqueeze calls that look a bit unusual to me.

**Comment by YvooneYIN** [comment]
> Hi! Have you seen our [official usage example](https://github.com/yandex-research/rtdl-revisiting-models/blob/main/package/example.ipynb)? I recommend trying to reproduce it without any changes. If it runs successfully, then it can be a good starting point for debugging your code. For example, I see that your code uses `sigmoid` + `BCELoss`, which is less numerically stable than using `BCEWithLogitsLoss`, and there is a bunch of squeeze/unsqueeze calls that look a bit unusual to me.

Thanks so much to your quick and helpful reply!
Sorry for my missing of such a clear example, at this time I am trying for the official example and hope it works!

By the way, I noticed that the FTTransformer framework appeares in several repositories, including [rtdl](https://github.com/yandex-research/rtdl), [tabular-dl-tabr](https://github.com/yandex-research/tabular-dl-tabr), [rtdl-num-embeddings](https://github.com/yandex-research/rtdl-num-embeddings) and this [rtdl-revisiting-models](https://github.com/yandex-research/rtdl-revisiting-models). 

I am a little confused about if there are any differences among them? 

Or could you give me any advices about which version should be use? Thanks!

**Comment by YvooneYIN** [comment]
Hi! I followed the example and it works well! Thanks so much !!!
When I have a closer exploration, I find out that the key is to choose the right combination of data preprocess method and optimizer! Standardization with StandardScaler + Adam works but SGD fails...
Anyway, the example helps a lot! The last question is the various FTTransformer in these repositories, could you please give me some instructions? Thanks!!!!

**Comment by Yura52** [comment]
Hi, glad to know that you resolved the issue!

The recommended implementation of FT-Transformer is [the official package](https://github.com/yandex-research/rtdl-revisiting-models/tree/main/package). Regarding other repositories:
- The [rtdl](https://github.com/yandex-research/rtdl) repository now serves as a meta repository with project links, but the `rtdl` package itself is deprecated.
- The [rtdl-num-embeddings](https://github.com/yandex-research/rtdl-num-embeddings) and [tabular-dl-tabr](https://github.com/yandex-research/tabular-dl-tabr) are implementations of other papers. Of course, as a coauthor of those papers, I can assure that FT-Transformer there is also correct :) However, strictly speaking, they should be treated as third party papers that copied FT-Transformer from official sources and adjusted it for their needs.

**Comment by YvooneYIN** [comment]
Hello, I have another question these days. I am looking at the [feature importance calculating](https://github.com/yandex-research/rtdl-revisiting-models/issues/2), and I understand the process of the code, but just a little confusion about: which one should this calculation be done: training dataset or on the testing data?
The article suggests "For the i-th sample, we calculate the average attention map pi for the [CLS] token from Transformer’s forward pass." Does this means I should take another forward pass on the test data? By the way, this seems exactly [the way others did](https://github.com/vcherepanova/tabular-feature-selection/blob/main/deep_tabular/utils/testing.py). Could you please give me some advices? Many thanks!

**Comment by YvooneYIN** [comment]
I've been reflecting on the questions I asked previously and have come to some new insights that I'd like to share and seek further opinions on:

1. I now think that evaluating these on test data might indeed be more suitable. This is because these metrics are calculated based on the values of the inputs and the model, and the attention probabilities dynamically change across different inputs.  So, to prevent overfitting and to better understand the model's decision-making with new, unseen data, it is more suitable to choose the test data to perform the calculation.

2. It seems that different test data can result in varying importance of features, which might be inevitable. I am now considering whether it is advisable to use multiple datasets from the same classification problem to identify a robust set of important features. What are your thoughts on this approach?

3. I realized that during the evaluation phase in PyTorch, an implicit forward pass is performed. It seems practical to output the attention probabilities during this phase to simultaneously obtain feature importance and other metrics from the same model run. 

I'd appreciate any feedback or further insights into these thoughts!

Thank you!

**Comment by Yura52** [comment]
Hello @YvooneYIN , I am sorry for the slow reply.

Without any additional context, I definitely think that feature importances should be calculated using _training_ data. Some quick thoughts:

1. This calculation is no different from any other calculation (e.g. model training) in a sense that, in general, one does _not_ have access to future "test" requests. For example, let's consider a case when feature importances are used to identify uninformative features and retrain the model with only important features. Then using test data for calculating feature importances means using future unknown data to improve a model on this future unknown data, which is not fair.

2. As you already mentioned, feature importances computed on a test dataset depend on the choice of a test dataset. This sounds counter-intuitive to me :)

3. Regarding other projects, as far as I remember, all decision-tree-based algorithms compute feature importances on training data without test data (XGBoost, CatBoost, algorithms from sklearn, ...).

That said, there may be some special cases where "future unknown data" can actually be known, for example, test data in Kaggle competitions.

P.S. In future, I suggest opening a new issue if a new question is on a different topic. This will make it easier for others to discover relevant discussions.

**Comment by YvooneYIN** [comment]
> Hello @YvooneYIN , I am sorry for the slow reply.
> 
> Without any additional context, I definitely think that feature importances should be calculated using _training_ data. Some quick thoughts:
> 
> 1. This calculation is no different from any other calculation (e.g. model training) in a sense that, in general, one does _not_ have access to future "test" requests. For example, let's consider a case when feature importances are used to identify uninformative features and retrain the model with only important features. Then using test data for calculating feature importances means using future unknown data to improve a model on this future unknown data, which is not fair.
> 2. As you already mentioned, feature importances computed on a test dataset depend on the choice of a test dataset. This sounds counter-intuitive to me :)
> 3. Regarding other projects, as far as I remember, all decision-tree-based algorithms compute feature importances on training data without test data (XGBoost, CatBoost, algorithms from sklearn, ...).
> 
> That said, there may be some special cases where "future unknown data" can actually be known, for example, test data in Kaggle competitions.
> 
> P.S. In future, I suggest opening a new issue if a new question is on a different topic. This will make it easier for others to discover relevant discussions.

Hi, I am totally agree with you that when using feature importance as a  feature selection method, it should be done on the training data to avoid information leak from test data.
However, what if I only treat this process as a way to understand the model decision?

**Comment by Yura52** [comment]
> However, what if I only treat this process as a way to understand the model decision?

Collecting and analysing attention maps on inference is definitely a valid thing to do! It just seems to me that the result of this procedure cannot be called "feature importances" in the usual sense of this term, as I understand it.
